File: plugin/src/main/java/de/s42/tradecycle/common/VersionCheckTask.java
package de.s42.tradecycle.common;

import lombok.AllArgsConstructor;
import org.bukkit.scheduler.BukkitRunnable;

import java.io.InputStream;
import java.net.URI;
import java.net.URL;
import java.util.Scanner;
import java.util.function.Consumer;

@AllArgsConstructor
public class VersionCheckTask extends BukkitRunnable {

    public static final String RESOURCE = "https://api.modrinth.com/v2/project/GpyBM69F/version?version_type=release";
    public static final String DIRECT_DOWNLOAD = "https://modrinth.com/plugin/tradecycle#download";

    private final Consumer<ComparableVersion> comparable;

    @Override
    public void run() {
        try {
            URL url = URI.create(RESOURCE).toURL();
            try (InputStream result = url.openStream()) {
                Scanner scanner = new Scanner(result);
                String json = scanner.useDelimiter("\\A").next();

                int idx = json.indexOf("\"version_number\":");
                if (idx != -1) {
                    int start = json.indexOf("\"", idx + 17) + 1;
                    int end = json.indexOf("\"", start);
                    String latestVersion = json.substring(start, end);
                    comparable.accept(new ComparableVersion(latestVersion));
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: core/src/main/kotlin/tr/s42/core/common/VersionCheckTask.kt
package net.cutecraft.core.common

import org.bukkit.scheduler.BukkitRunnable
import java.net.URI
import java.util.*

class VersionCheckTask(
    val comparable: (ComparableVersion) -> Unit
) : BukkitRunnable() {

    companion object {
        const val RESOURCE = "https://api.spigotmc.org/legacy/update.php?resource=122805"
    }

    override fun run() {
        try {
            val url = URI(RESOURCE).toURL()
            url.openStream().use { result ->
                val scanner = Scanner(result)
                comparable(ComparableVersion(scanner.next()))
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

}